pkg/providers/vmware/client/client.go

```diff
@@ -0,0 +1,145 @@
+package client
+
+import (
+	"context"
+	"github.com/vmware/govmomi/object"
+	"github.com/vmware/govmomi/session"
+	"github.com/vmware/govmomi/vim25"
+	"github.com/vmware/govmomi/vim25/mo"
+	"github.com/vmware/govmomi/vim25/soap"
+	"github.com/vmware/govmomi/vim25/types"
+	"net/url"
+	"time"
+)
+
+// timeout value in seconds for vmware api requests
+const timeout = 30 * time.Second
+
+// RichOvirtClient is responsible for retrieving VM data from oVirt API
+type RichVmwareClient struct {
+	client         *vim25.Client
+	url            *url.URL
+	sessionManager *session.Manager
+}
+
+// NewRichVMwareClient creates new, connected rich vmware client. After it is no longer needed, call Close().
+func NewRichVMWareClient(apiUrl, username, password string, thumbprint string) (*RichVmwareClient, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	u, err := url.Parse(apiUrl)
+	if err != nil {
+		return nil, err
+	}
+	if u.User == nil {
+		u.User = url.UserPassword(username, password)
+	}
+
+	soapClient := soap.NewClient(u, false)
+	soapClient.SetThumbprint(u.Host, thumbprint)
+	vimClient, err := vim25.NewClient(ctx, soapClient)
+	if err != nil {
+		return nil, err
+	}
+
+	sessionManager := session.NewManager(vimClient)
+	err = sessionManager.Login(ctx, u.User)
+	if err != nil {
+		return nil, err
+	}
+
+	vmwareClient := RichVmwareClient{
+		client:         vimClient,
+		url:            u,
+		sessionManager: sessionManager,
+	}
+	return &vmwareClient, nil
+}
+
+func (r RichVmwareClient) GetVM(id *string, _ *string, _ *string, _ *string) (interface{}, error) {
+	return r.getVM(*id), nil
```

Comment:
From the API standpoint VM ID is optional and can be nil -  in that case this line will cause panic, if it is not set to some value in the provider.